server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/builder/OffHeapStarTreeBuilder.java

```diff
@@ -0,0 +1,896 @@
+            output.writeByte(bitSet);
+            numBytes += Byte.BYTES;
+        }
+        return numBytes;
+    }
+
+    /**
+     * Reads the star tree document from file with given offset
+     *
+     * @param input   RandomAccessInput
+     * @param offset  Offset in the file
+     * @param shouldReadAggregatedDocs boolean to indicate if aggregated star tree docs should be read
+     * @return StarTreeDocument
+     * @throws IOException IOException in case of I/O errors
+     */
+    private StarTreeDocument readStarTreeDocument(RandomAccessInput input, long offset, boolean shouldReadAggregatedDocs)
+        throws IOException {
+        int dimSize = starTreeField.getDimensionsOrder().size();
+        Long[] dimensions = new Long[dimSize];
+        offset = readDimensions(dimensions, input, offset);
+
+        int numMetrics = getNumMetrics();
+        Object[] metrics = new Object[numMetrics];
+        offset = readMetrics(input, offset, numMetrics, metrics, shouldReadAggregatedDocs);
+
+        return new StarTreeDocument(dimensions, metrics);
+    }
+
+    /**
+     * Read dimensions from file
+     */
+    long readDimensions(Long[] dimensions, RandomAccessInput input, long offset) throws IOException {
+        for (int i = 0; i < dimensions.length; i++) {
+            try {
+                dimensions[i] = input.readLong(offset);
+            } catch (Exception e) {
+                logger.error("Error reading dimension value at offset {} for dimension {}", offset, i);
+                throw e;
+            }
+            offset += Long.BYTES;
+        }
+        offset += setNullValuesForDimensions(input, offset, dimensions);
+        return offset;
+    }
+
+    /**
+     * Read star tree metrics from file
+     */
+    private long readMetrics(RandomAccessInput input, long offset, int numMetrics, Object[] metrics, boolean shouldReadAggregatedDocs)
+        throws IOException {
+        for (int i = 0; i < numMetrics; i++) {
+            switch (metricAggregatorInfos.get(i).getValueAggregators().getAggregatedValueType()) {
+                case LONG:
+                    metrics[i] = input.readLong(offset);
+                    offset += Long.BYTES;
+                    break;
+                case DOUBLE:
+                    long val = input.readLong(offset);
+                    if (shouldReadAggregatedDocs) {
+                        metrics[i] = StarTreeNumericTypeConverters.sortableLongtoDouble(val);
+                    } else {
+                        metrics[i] = val;
+                    }
+                    offset += Long.BYTES;
+                    break;
+                default:
+                    throw new IllegalStateException("Unsupported metric type");
+            }
+        }
+        offset += setNullForMetrics(input, offset, numMetrics, metrics);
+        return offset;
+    }
+
+    private int getNumMetrics() {
+        int numMetrics = 0;
+        for (Metric metric : starTreeField.getMetrics()) {
```

Comment:
Can this change? We should be able to cache this within the constructor only, right?